src/libraries/System.Net.WebSockets/tests/WebSocketCloseTests.cs

```diff
@@ -0,0 +1,89 @@
+
+        public WebSocketCloseTests()
+        {
+            if (!Debugger.IsAttached)
+            {
+                _cancellation = new CancellationTokenSource(TimeSpan.FromSeconds(5));
+            }
+        }
+
+        public CancellationToken CancellationToken => _cancellation?.Token ?? default;
+
+        public static object[][] CloseStatuses = {
+            new object[] { WebSocketCloseStatus.EndpointUnavailable },
+            new object[] { WebSocketCloseStatus.InternalServerError },
+            new object[] { WebSocketCloseStatus.InvalidMessageType},
+            new object[] { WebSocketCloseStatus.InvalidPayloadData },
+            new object[] { WebSocketCloseStatus.MandatoryExtension },
+            new object[] { WebSocketCloseStatus.MessageTooBig },
+            new object[] { WebSocketCloseStatus.NormalClosure },
+            new object[] { WebSocketCloseStatus.PolicyViolation },
+            new object[] { WebSocketCloseStatus.ProtocolError },
+            new object[] { (WebSocketCloseStatus)1012 },  // ServiceRestart indicates that the server / service is restarting.
+            new object[] { (WebSocketCloseStatus)1013 },  // TryAgainLater indicates that a temporary server condition forced blocking the client's request.
+            new object[] { (WebSocketCloseStatus)1014 },  // BadGateway indicates that the server acting as gateway received an invalid response
+        };
+
+        [Theory]
+        [MemberData(nameof(CloseStatuses))]
+        public void WebSocketReceiveResult_WebSocketCloseStatus_Roundtrip(WebSocketCloseStatus closeStatus)
+        {
+            string serverMessage = "closeStatus " + closeStatus.ToString();
+            WebSocketReceiveResult wsrr = new WebSocketReceiveResult(42, WebSocketMessageType.Close, true, closeStatus, serverMessage);
+            Assert.Equal(42, wsrr.Count);
+            Assert.Equal(closeStatus, wsrr.CloseStatus);
+            Assert.Equal(serverMessage, wsrr.CloseStatusDescription);
+        }
+
+        [Theory]
+        [MemberData(nameof(CloseStatuses))]
+        public async Task ReceiveAsync_ValidCloseStatus_Success(WebSocketCloseStatus closeStatus)
+        {
+            byte[] receiveBuffer = new byte[1024];
+            WebSocketTestStream stream = new();
+            Encoding encoding = Encoding.UTF8;
+
+            using (WebSocket server = WebSocket.CreateFromStream(stream, true, null, TimeSpan.FromSeconds(3)))
+            using (WebSocket client = WebSocket.CreateFromStream(stream.Remote, false, null, TimeSpan.FromSeconds(3)))
+            {
+                Assert.NotNull(server);
+                Assert.NotNull(client);
+
+                // send something
+                string hello = "Testing " + closeStatus.ToString();
+                byte[] sendBytes = encoding.GetBytes(hello);
+                await server.SendAsync(sendBytes.AsMemory(), WebSocketMessageType.Text, WebSocketMessageFlags.DisableCompression, CancellationToken);
```

Comment:
```suggestion
                string hello = "Testing " + closeStatus.ToString();
                byte[] sendBytes = encoding.GetBytes(hello);
                await server.SendAsync(sendBytes.AsMemory(), WebSocketMessageType.Text, WebSocketMessageFlags.EndOfMessage, CancellationToken);
```

Reply:
I thing we need the `DisableCompression` to ensure we don't get compressed data out (I borrowed that from elsewhere).

As for the `EndOfMessage`, I _intentionally_ switch away from that being sent that way so we are testing a path more reflective of the real world, where we're sending and sending and then hang-up. If that's a bad understanding (and it easliy can be, this is NOT my area), I would happily apply this change, but I would rather it still have the compression off like it was originally: 
```csharp
await server.SendAsync(sendBytes.AsMemory(),
                       WebSocketMessageType.Text,
                       WebSocketMessageFlags.DisableCompression | WebSocketMessageFlags.EndOfMessage,
                       CancellationToken);
```

Reply:
Compression is opt-in, so unless you created your WebSocket instance with [`DangerousDeflateOptions`](https://learn.microsoft.com/en-us/dotnet/api/system.net.websockets.websocketcreationoptions.dangerousdeflateoptions?view=net-7.0#system-net-websockets-websocketcreationoptions-dangerousdeflateoptions) creation option (in "real world" it would start with both client *and* server opting in compression and negotiating it during initial handshake) -- this flag would be ignored, as the compression is off.

Also, re "we don't get compressed data out" -- even if compression IS enabled, it is completely transparent for end-users, it will get automatically decompressed upon arrival.

As for EndOfMessage, I'm fine either way.

Reply:
I told you it wasn't my area :) 

I'll update that when I rebase everything... in process now.